### ISSUE TITLE
`get_run_history_tool`: silent empty result for keys logged in different `log()` calls, and lossy behavior when `min_step`/`max_step` is set

### ISSUE THREAD

**Issue by zacwitte** [wandb/wandb-mcp-server#136]
`get_run_history_tool` returns complete, correct data when all requested keys share a logging cadence and no step range is set. Outside that case it fails in two ways that are indistinguishable from "this metric does not exist". Neither emits an error or a warning.

Run shape used below: 154,196 history rows; `train/loss` logged every ~401 steps (375 rows total); `optimizer_step` logged on a denser cadence in a **separate** `log()` call.

## Issue 1 — keys logged in different `log()` calls return zero rows, silently

AND semantics: a row is returned only if it carries every requested key. Keys never written in the same `log()` call have no such row, so the result is empty.

```jsonc
// Same log() call - correct
{"keys": ["train/loss", "train/aux_loss"], "samples": 2000}
// -> 375 rows, both keys on every row

// Different log() calls - silently empty
{"keys": ["train/loss", "optimizer_step"], "samples": 2000}
// -> {"rows": [], "sampled_points": 0, "keys_returned": []}

// Either key alone - full history
{"keys": ["train/loss"], "samples": 2000}      // -> 375 rows
{"keys": ["optimizer_step"], "samples": 2000}  // -> full series
```

**Expected:** the union of rows, each carrying whichever requested keys exist at that step; or an explicit diagnostic.

**Actual:** `rows: []`, `keys_returned: []`, no error.

Which metrics share a `log()` call is an implementation detail of the training loop that the caller cannot see. Any request mixing a loss with a step counter, a learning rate, or a throughput metric hits this.

**Suggested fixes**, in preference order:

1. Return the **union** of rows; let each row carry whichever requested keys exist at that step.
2. Return **one series per key**, so cadence differences cannot interact at all.
3. At minimum, when the result is empty but the keys individually have data, say so in the response, with per-key row counts.

A `join` parameter (`"outer"` default, `"inner"` opt-in) would fix this while preserving today's semantics for anyone depending on them.

## Issue 2 — `min_step`/`max_step` returns mostly empty rows and drops values

With a step range set, the server appears to sample raw history rows across the range and then project them onto the requested keys, rather than restricting to rows that carry those keys. This is lossy even for a **single** key, so it is independent of Issue 1.

The window `[0, 6000]` contains exactly 14 `train/loss` rows.

```jsonc
// No range - complete
{"keys": ["train/loss"], "samples": 2000}
// -> 375 rows, every row carries the key, zero empty rows

// Range added - 2000 rows, 8 useful ones
{"keys": ["train/loss"], "samples": 2000, "max_step": 6000}
// -> only 8 of the 14 in-window values present; the rest are {"_step": N} shells

// Two cross-cadence keys + range - no row has both
{"keys": ["train/loss", "optimizer_step"], "samples": 2000, "max_step": 6000}
// -> 2000 rows: 5 carry train/loss, 1912 carry optimizer_step, 0 carry both, 83 carry neither
```

**Expected:** a step range narrows the same result set the unranged call returns, same rows and fewer of them, with `samples` budgeting rows that carry requested data.

**Actual:** lossy (8 of 14 values) and wasteful (2,000 rows to deliver 8 values).

In the last case `keys_returned` lists both keys, which reads as success, but no row carries both and there is no way to join them. At smaller `samples` the sparser key disappears from `keys_returned` entirely, so whether a key appears to exist depends on the sampling budget.

**Suggested fix:** apply the step range as a filter on the key-bearing rows, so ranged and unranged requests differ only in extent, and spend the `samples` budget on rows carrying at least one requested key.

## Impact

Together these make the documented "plot the loss curve for run X" path unreliable for any run whose metrics are logged on more than one cadence, which is most real training runs. The workaround (one request per logging group, no step range, `samples` above the row count, join on `_step` client-side) works well, but it has to be discovered by experiment, because nothing in the response indicates that the obvious call returned incomplete or empty data.

Related discoverability fix: have `probe_project_tool` report which keys co-occur (their logging groups) and each key's row count, so callers can size `samples` and group their requests without trial and error.

**Comment by ash0ts** [comment]
The fix is merged into `staging/0.4.0` at `774d5fe119d42c936c7d1221b5a480e8d50e5eaf` via #137 and has passed live Cloud Run staging validation.

Validated through the official MCP `ClientSession` against run `h0fm5qp5`:

- unranged `loss` + `eval/loss`, `samples=500`: 500 nonempty outer-union rows, 250 values per key
- range `0..599`, `samples=500`: exactly 400 useful rows, 200 values per key, `rows_scanned=600`, zero empty shells
- known values, missing-key metadata, session continuity, and response safety all passed

Evidence: [staging workflow #30944209088](https://github.com/wandb/wandb-mcp-server-test/actions/runs/30944209088). Release PR #117 is ready for human review but intentionally remains unmerged, so this issue should stay open until v0.4 lands on `main`.
